src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.Linux.cs

```diff
@@ -328,6 +332,45 @@ protected ulong ClearRegister
             set {  *(ulong*)(_registerViewPointer->GPCLR) = value; }
         }
 
+        /// <summary>
+        /// Returns the peripheral base address on the CPU bus of the raspberry pi based on the ranges set within the device tree.
+        /// </summary>
+        /// <remarks>
+        /// The range examined in this method is essentially a mapping between where the peripheral base address on the videocore bus and its
+        /// address on the cpu bus. The return value is 32bit (is in the first 4GB) even on 64 bit operating systems (debian / ubuntu tested) but may change in the future
+        /// This method is based on bcm_host_get_peripheral_address() in libbcm_host which may not exist in all linux distributions.
+        /// </remarks>
+        /// <returns>This returns the peripheral base address as a 32 bit address or 0xFFFFFFFF when in error.</returns>
+        private uint GetPeripheralBaseAddress()
+        {
+            uint cpuBusPeripheralBaseAddress = 0xFFFFFFF;
+            uint vcBusPeripheralBaseAddress;
+
+            using (BinaryReader rdr = new BinaryReader(File.Open(DeviceTreeRanges, FileMode.Open, FileAccess.Read)))
+            {
+                // get the Peripheral Base Address on the VC bus from the device tree this is to be used to verify that
+                // the right thing is being read and should always be 0x7E000000
+                vcBusPeripheralBaseAddress = BinaryPrimitives.ReadUInt32BigEndian(rdr.ReadBytes(4));
+
+                // get the Peripheral Base Address on the CPU bus from the device tree.
+                cpuBusPeripheralBaseAddress = BinaryPrimitives.ReadUInt32BigEndian(rdr.ReadBytes(4));
+
+                // if the CPU bus Peripheral Base Address is 0 then assume that this is a 64 bit address and so read the next 32 bits.
+                if (cpuBusPeripheralBaseAddress == 0)
+                {
+                    cpuBusPeripheralBaseAddress = BinaryPrimitives.ReadUInt32BigEndian(rdr.ReadBytes(4));
+                }
+
+                // if the address values don't fall withing known values then assume an error
+                if (vcBusPeripheralBaseAddress != 0x7E000000 || !(cpuBusPeripheralBaseAddress == 0x20000000 || cpuBusPeripheralBaseAddress == 0x3F000000 || cpuBusPeripheralBaseAddress == 0xFE000000))
```

Comment:
would be useful to add some comment to where those constants come from (i.e. spec page or some link)

Reply:
added a link to the DTS repository in GitHub. Now have named constants that indicate which chipset the values are associated with,